### PR TITLE
Apply USGS styling to xml publications

### PR DESCRIPTION
**Pull request by ssoper-usgs**
Before making a pull request
----------------------------

- [x] Make sure all tests run
- [ ] Update the changelog appropriately

Title
-----------
JIRA:
https://internal.cida.usgs.gov/jira/browse/PUBSTWO-1676
https://internal.cida.usgs.gov/jira/browse/PUBSTWO-1678

Description
-----------
Prototype transformation using BeautifulSoup4 to apply USGS styling to xml publications.

After making a pull request
---------------------------
- [x] If appropriate, put the link to the PR in the JIRA ticket
- [x] Assign someone to review unless the change is trivial

### CODE REVIEW

File: server/pubs_ui/pubswh/utils.py [hunk 1]
```diff
@@ -311,6 +311,137 @@ def pull_feed(feed_url):
     return post
 
 
+def transform_html(html, image_url):
```
Comment (ssoper-usgs) [inline server/pubs_ui/pubswh/utils.py]: Seems like this function could use a bit of breakdown: https://app.codacy.com/manual/usgs_wma_dev/pubswh-ui/pullRequest?prid=4169950 - any thoughts on how to divide it up?  My initial thought is to create smaller functions for building the citations table and the image figures.

**Comment by mbucknell** [comment]
Can you break it down by component? Fixing styling for figures, fixing it for tables, fixing it for lists?

I haven't looked at this in great detail yet.

File: server/pubs_ui/pubswh/utils.py [hunk 2]
```diff
@@ -311,6 +311,137 @@ def pull_feed(feed_url):
...
+
+    ref_list_rows = ref_list.findAll('div', 'row')
+    for ref_list_row in ref_list_rows:
+
+        # add a new row to the citation table
+        row = soup.new_tag('tr')
+        citation_table.append(row)
+
+        # add a data cell for ref-label stuff
+        label_td = soup.new_tag('td')
```
Comment (mbucknell) [inline server/pubs_ui/pubswh/utils.py]: You might be able to make this a little easier to read by not declaring so many single use variables.So for instance is line 351 necessary? Or could you do label_td.append(ref_label.find('a'))

Comment (ssoper-usgs) [reply server/pubs_ui/pubswh/utils.py]: Yeah that makes sense, I'll get rid of a bunch of the single use variables first before breaking down this function further.

File: server/pubs_ui/pubswh/xml_transformations.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    citation_table = soup.new_tag('table')
+    citation_table['id'] = 'references-cited'
+
+    for ref_list_row in references.findAll('div', 'row'):
+        # add a new row to the citation table
+        row = soup.new_tag('tr')
+        citation_table.append(row)
+
+        # add a data cell for ref-label content
+        label_td = soup.new_tag('td')
```
Comment (mbucknell) [inline server/pubs_ui/pubswh/xml_transformations.py]: I would really encourage the elimination of these one time variables. So lines 77 and 78 would become:
```
row.append(soup.new_tag('td'). 
```
Not only is it more succient, I belive it is easier to see what is happening. There are additional ones in this function and others.

File: server/pubs_ui/pubswh/tests/test_xml_transformations.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        expected_string = str(soup.find('body'))
+        expected_no_whitespace = "".join(expected_string.split())
+
+        actual_string = str(transform_xml_full(app.config['HTML_ENDPOINT'], app.config['SPN_IMAGE_URL']))
+        actual_no_whitespace = "".join(actual_string.split())
+
+        # string comparison of the soup output with all whitespace removed
+        self.assertEqual(actual_no_whitespace, expected_no_whitespace)
+
+    def test_does_the_transform_produce_a_citation_table(self):
```
Comment (mbucknell) [inline server/pubs_ui/pubswh/tests/test_xml_transformations.py]: Typically, we put each function's tests in a separate TestCase class. So I would expect to see a GetCitationTableTestCase for instance and so on. This allows us to more easily find where specific code is tested and prevents enormous test cases which can be tricky when refactoring.

File: server/config.py [hunk 5]
```diff
@@ -109,4 +109,13 @@
 # Set to False when running the development server on https
 SECURE_COOKIES = True
 
+# HTML_ENDPOINT will eventually be a url for pulling html from pubs-services
+sample_html_absolute_path = os.path.abspath(os.path.dirname(__file__))
+sample_html_relative_path = os.path.join(sample_html_absolute_path, "pubs_ui/pubswh/tests/data/sample_html.html")
+with open(sample_html_relative_path) as f:
+    HTML_ENDPOINT = f.read()
```
[the remark this answers is not in the document]
Comment (ssoper-usgs) [reply server/config.py]: Good point! Change on the way!

File: server/pubs_ui/pubswh/tests/test_xml_transformations.py (hunk 4 above)
Comment (ssoper-usgs) [reply server/pubs_ui/pubswh/tests/test_xml_transformations.py]: Excellent, will implement this.

File: server/pubs_ui/pubswh/xml_transformations.py (hunk 3 above)
Comment (ssoper-usgs) [reply server/pubs_ui/pubswh/xml_transformations.py]: I do use `label_td` down on lines 82 and 83.  However!  I can still implement the change you suggested and then, on lines 82-83 and elsewhere, I could use a different and perhaps more readable approach used in BeautifulSoup like this: `soup.table.tr.td.append(ref_label.find('span', 'generated'))` and so on.

Comment (mbucknell) [reply server/pubs_ui/pubswh/xml_transformations.py]: That sounds great!

**Pull request merged**